### DISCUSSION TITLE
Custom authentication

### DISCUSSION THREAD

**Discussion by nurmukhametdaniyar** [supabase/supabase#2089] [Questions] [answered]
I have been trying to implement a custom Auth Provider (Twilio) with Supabase, but couldn't find a way how to do that. There is discussion #417 where it states that it is possible, but it wasn't described how it can be done and how we store users in Supabase.

What is the workflow for the custom authentication in Supabase?

**Comment by silentworks** [comment]
Is this a client-side application or a server-side application? because depending on which it is the workflow will be different, I don't think Twilio has a client-side auth library, so I don't think it will be possible to use it client-side.

**Reply by nurmukhametdaniyar** [reply]
Ok, let's suppose that I got JWT token, is there a way to store and authenticate with custom tokens in Supabase?

**Reply by cspecter** [reply]
I would like to know how to do this as well. We use Firebase for authentication as we need phone auth for our app. But I don't see how to add or authenticate the Firebase JWT in Supabase.

We have tried something like this:

```javascript
const token = await firebase.auth().currentUser.getIdToken()
supabase.auth.setAuth(token)
```

But we get this error: `invalid algorithm` on our calls.

**Reply by ashleyconnor** [reply]
That's not going to work because you're hitting Supabase with the Firebase JWT. To do custom auth you are going to need to run custom code somewhere, say on an API or serverless function, to sign in against the third-party authentication system.

To clarify, I think [this comment](https://github.com/supabase/supabase/discussions/417#discussioncomment-229555) suggests that once you've validated a user using for example Auth0, you need to sign a JWT in your API/serverless function and use that to initialize the Supabase client so that RLS, using your policies, works correctly.

You can sign the Supabase JWT using the same secret Supabase uses, which can be found by running `show app.settings.jwt_secret;` (never expose this to your users) on the SQL tab. Other signing settings will also need to match Supabase. The `HS256` algorithm for example.

**Reply by cspecter** [reply]
So I am currently decoding the JWT from Firebase, then re-encoding it with the Supabase JWT secret. This lets me create a Supabase session. But now it's telling me that the user isn't in the `users` table when I try to upload a file, which it isn't, since the app is authenticated via Firebase.

In storage I have a public bucket and this policy `(role() = 'authenticated':: text)`.

I'm not sure how to proceed from here. Ideally we would have a working sample to look at for using third-party auth with Row Level security.

**Reply by ashleyconnor** [reply]
The `storage.objects` table has a column `owner` of type `uuid`. There's a foreign key constraint which is why it fails if you're not using Supabase's auth.

You might be able to drop that constraint but I'm not 100% sure it will populate the column if you do.

```
Foreign-key constraints:
    ...
    "objects_owner_fkey" FOREIGN KEY (owner) REFERENCES auth.users(id)
```

**Reply by cspecter** [reply]
Hmmm, so not really set up to handle third-party auth. Maybe we'll have to look at uploading files from a serverless function and not use RLS. Supabase really needs a nodeJs admin module like Firebase.

**Reply by ashleyconnor** [reply]
You could also duplicate the user in Supabase using a cloud function from the Firebase side.

**Reply by cspecter** [reply]
We are looking at that. Not ideal though. Too bad Supabase doesn't have phone auth + text yet...

**Reply by inian** [reply]
Phone Auth is a WIP here https://github.com/supabase/gotrue/pull/115

**Reply by silentworks** [reply]
Phone Auth is now live. You can read more about it in this blog post https://supabase.io/blog/2021/07/28/supabase-auth-passwordless-sms-login

**Comment by kangmingtay** [comment, marked as answer]
Hey @nurmukhametdaniyar, we now support phone auth with twilio! Check out our [blog post](https://supabase.io/blog/2021/07/28/supabase-auth-passwordless-sms-login) or the [docs](https://supabase.io/docs/guides/auth/auth-twilio) to get started!
